### PR TITLE
stats

### PR DESCRIPTION
**Pull request by terencechain**

### CODE REVIEW

File: src/app/stats/stats.component.scss [hunk 1]
```diff
@@ -0,0 +1,23 @@
+@import '../../styles/variables';
+
+.charts {
+    position: fixed;
+    top: 0;
+    left: 0;
+    padding: 120px 20px;
+    height: 100% !important;
+    width: 100% !important;
+    background-color: white;
```
Comment (LinoleumKnife) [inline src/app/stats/stats.component.scss:10]:
```suggestion
    background-color: $white;
```

File: src/app/stats/stats.component.ts [hunk 2]
```diff
@@ -7,9 +34,78 @@ import { Component, OnInit } from '@angular/core';
...
+        existingParentBlocks[blockRoot] = true;
+
+        let participatedVotesString = Math.floor(node.getParticipatedVotes() / MAX_DEPOSIT_AMOUNT).toString();
+        if (participatedVotesString.indexOf('.') === -1) {
+          participatedVotesString += '.';
+        }
+        while (participatedVotesString.length < participatedVotesString.indexOf('.') + 15) {
+          participatedVotesString += '0';
+        }
+        let totalVotesString = Math.floor(node.getParticipatedVotes() / MAX_DEPOSIT_AMOUNT).toString();
```
Comment (LinoleumKnife) [inline src/app/stats/stats.component.ts:74]:
```suggestion
        let totalVotesString = Math.floor(node.getTotalVotes() / MAX_DEPOSIT_AMOUNT).toString();
```

I'm assuming. Didn't follow the logic thoroughly, but the logic for `participatedVotesString` and `totalVotesString` is currently identical

File: src/app/stats/stats.component.ts [hunk 3]
```diff
@@ -1,4 +1,31 @@
...
+}
+
+interface GraphEdge {
+  id: string | number;
+  source: string;
+  target: string;
+}
+
+const GENESIS_SLOT = '9223372036854775808';
+const MAX_DEPOSIT_AMOUNT = 3200000000;
```
Comment (LinoleumKnife) [inline src/app/stats/stats.component.ts:27]: What is special about this number versus `environment.depositAmount`?

File: src/app/stats/stats.component.ts [hunk 4]
```diff
@@ -7,9 +34,78 @@ import { Component, OnInit } from '@angular/core';
...
+          participatedVotes: participatedVotesString,
+          totalVotes: totalVotesString,
+          numAttestations: node.getBlock().getBody().getAttestationsList().length,
+          numDeposits: node.getBlock().getBody().getDepositsList().length,
+          slot: (new BigNumber(node.getBlock().getSlot()).minus(new BigNumber(GENESIS_SLOT))).toString(),
+        };
+      });
+
+      for (let i = 0; i < this.nodes.length; i++) {
+        const node = tree[i];
```
Comment (LinoleumKnife) [inline src/app/stats/stats.component.ts:94]:
```suggestion
        const node = this.nodes[i];
```
?

Also, any reason why not using
`for (const item of items)`

File: src/app/stats/stats.component.scss [hunk 5]
```diff
@@ -0,0 +1,23 @@
+@import '../../styles/variables';
+
+.charts {
+    position: fixed;
+    top: 0;
+    left: 0;
+    padding: 120px 20px;
+    height: 100% !important;
```
Comment (prestonvanloon) [inline src/app/stats/stats.component.scss:8]: No `!important`

File: src/app/stats/stats.component.scss [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    width: 100% !important;
```
Comment (prestonvanloon) [inline src/app/stats/stats.component.scss:9]: No `!important`

File: src/app/stats/stats.component.ts (hunk 3 above)
Comment (prestonvanloon) [reply src/app/stats/stats.component.ts:27]: +1

File: src/app/stats/stats.component.ts [hunk 7]
```diff
@@ -7,9 +34,78 @@ import { Component, OnInit } from '@angular/core';
...
+  // slotPoll() {
+  //   return interval(SLOT_DURATION_MILLISECONDS)
+  //     .pipe(
+  //       switchMap(() => {
+  //         this.progress.startProgress();
+  //         return this.blockService.getBlockTree();
+  //       }), 
+  //       // distinctUntilChanged(), // only emit and redraw graph if the value has actually changed
+  //     );
+  // }
```
Comment (prestonvanloon) [inline src/app/stats/stats.component.ts:57]: Remove commented block

File: src/app/stats/stats.component.ts [hunk 8]
```diff
@@ -7,9 +34,78 @@ import { Component, OnInit } from '@angular/core';
...
+          this.links.push({
+            id: blockRoot.slice(0, 3),
+            source: parentRoot,
+            target: blockRoot,
+          });
+        }
+      }
+      // this.update$.next(true);
+      // this.center$.next(true);
+      // this.progress.stopProgress();
```
Comment (prestonvanloon) [inline src/app/stats/stats.component.ts:108]: Remove commented block

File: src/app/stats/stats.component.ts (hunk 4 above)
Comment (prestonvanloon) [reply src/app/stats/stats.component.ts:94]: +1 to `for (const node of this.nodes)`

**Comment by nisdas** [comment]
@terencechain whats the status on this ?

**Comment by terencechain** [comment]
We can close it, the gRPC endpoint for blocktree is no longer viable. Let's discuss offline how whether we still want something like this

**Pull request closed without merging**